### ISSUE TITLE
bufio: document SplitFunc corner cases

### ISSUE THREAD

**Issue by yaxum62** [golang/go#25472]
The current document of `bufio.SplitFunc` does not cover all cases and left few details buried into implementations. Such as:

1.  A Scanner behave differently when a SplitFunc returns token with nil or an empty byte slice. [example](https://play.golang.org/p/ybw5tTtvdbk) 
1. A Scanner usually skip advanced bytes when returned token is nil. [example](https://play.golang.org/p/jneMm94tMF9)

Here I propose few changes to doc, clearly state those behavior. Such as:

1. If a SplitFunc returns a non-nil error, returned advance and token are ignored. 
1. If a SplitFunc returns a non-nil token, even if it is empty, Scanner will always yield, no matter what error a reader previously returns.
1. If a SplitFunc returns a nil token with non-zero advance, Scanner will skip those bytes.

**Comment by bcmills** [comment]
This issue seems two describe two separate issues: missing doc comments, and a requested change in behavior.

Please file those separately: documentation issues tend to be relatively easy (and uncontentious) to fix, whereas changes in behavior may require a more rigorous proposal, especially if they are not [compatible with Go 1](https://play.golang.org/p/PzHf02Fah3q).

**Comment by bradfitz** [comment]
@robpike, you want to clarify Scanner docs here?

**Comment by yaxum62** [comment]
The different behaviors between nil and empty slice is not that obvious since those two are generally interchangeable in golang. The difference should be clearly documented here if it is supported.

And also the feature "nil token skips bytes" are not always working as expected. In some case it will terminate the scan if the reader already hits a non-nil error. [example](https://play.golang.org/p/jneMm94tMF9). As @bcmills suggested the doc need to be fixed before we can fix the implementation.

**Comment by robpike** [comment]
I'm sorry but I still don't understand what you're after. If the reader hits an error, the scan always terminates. The token makes no difference. If you believe otherwise, please explain further.

I also disagree with the assertion that nil and empty slice are interchangeable. They're not in general.

**Comment by yaxum62** [comment]
If the reader returns error and split function returns nil token, scan
function skip everything still in the buffer, with certain buffer size. If
you check the example I posted

**Comment by pam4** [comment]
> The docs seem clear to me, unless you're arguing that a nil token is somehow special, which it is not.

It *is* special: returning `(n, nil, nil)` (with n > 0) from the `SplitFunc` doesn't produce any token, not even a nil token (either the `SplitFunc` is called again or `Scan` returns false).

Conversely returning `(n, []byte{}, nil)` *does* produce a zero-length token.

This distinction is undocumented; the docs only explicitly talks about the `(0, nil, nil)` case.

> If the reader returns error and split function returns nil token, scan function skip everything still in the buffer

I agree this is undocumented/confusing.
In other words: returning `(n, nil, nil)` (with n > 0) from the `SplitFunc` while `atEOF` is true always terminates the scan (even if n < len(data)).

And if `atEOF` is false, `Scan` immediately tries to read more data into the buffer before calling the `SplitFunc` again, which is another difference from the `(n, []byte{}, nil)` case.

**Comment by ianlancetaylor** [comment]
The docs describe what happens when a `SplitFunc` returns a `nil` token.  A `[]byte{}` is not a `nil` token.  As far as I can see the docs are accurate.  I don't see any reason to explicitly say anything about a non-`nil` slice with length zero.  It's non-`nil`, and everything else follows.

I think more clarification about returning `n, nil, nil`, where `n > 0`, might be reasonable.  I think it's there in the docs, which just say that `0, nil, nil` is a special case, but it takes spme thought to see that `n, nil, nil` is not a special case.

**Comment by pam4** [comment]
> The docs describe what happens when a SplitFunc returns a nil token.

As you said yourself, it doesn't.
It just talks about the `(0, nil, nil)` case.
If `(n, nil, nil)` also have a special effect, the docs should just say so, otherwise it is incomplete. There is no basis to infer it.

> As far as I can see the docs are accurate.

Accurate but incomplete.

> I don't see any reason to explicitly say anything about a non-nil slice with length zero.

I agree. That would be the normal case. I was just pointing out that the `(n, nil, nil)` case, being different, must be documented.

This only cover the first point in my previous message. There is also the second point: the condition that terminates the scan is also not documented.

**Comment by ianlancetaylor** [comment]
@pam4 Do you agree with the change I sent in https://golang.org/cl/118695 ?

I feel like the condition that terminates the scan is documented at `(*Scanner).Scan` and `ErrFinalToken`.

**Comment by ianlancetaylor** [comment]
Thanks, I added a comment about what happens if the `SplitFunc` returns a `nil` token when called with `atEOF` set to true.

**Comment by pam4** [comment]
Thanks.

> If the token is not nil, the Scanner returns it to the user.

I think this sentence is very effective to clarify my first point.

> If the token is nil, the Scanner reads more data and continues scanning;
> if there is no more data--if atEOF was true--the Scanner returns.

It is correct, but how about something like:

```
If the token is nil, the Scanner reads more data and continues scanning,
unless atEOF was true, in which case the scan stops (Scan returns false).
```

**Comment by pam4** [comment]
`(*Scanner).Scan`:
> It returns false when the scan stops, either by reaching the end of the input or an error.

I just discovered that I had a misconception about it, which was probably caused by the above sentence.

There are exactly 3 ways for the scan to stop. None of them could be really described as "by reaching the end of the input":

1) the `SplitFunc` returns error
2) the `SplitFunc` returns nil token while `atEOF` is true
3) returning more than `maxConsecutiveEmptyReads` tokens without advancing the input (panic)

I think it is easy to interpret the quoted sentence to mean that you can also stop the scan by just advancing the input to the end. Probably the OP was also confused by this, considering his examples.
